File: _tmp/test_style_copy.mjs
// Smoke-test addMissingStyles against a temp copy of the decks. No server.
import fs from "node:fs/promises";
import path from "node:path";
import os from "node:os";
import url from "node:url";

const REPO = path.resolve(url.fileURLToPath(import.meta.url), "../..");
const TMP = await fs.mkdtemp(path.join(os.tmpdir(), "slides-test-"));

async function cp(src, dst) {
  await fs.cp(src, dst, { recursive: true });
}

await cp(path.join(REPO, "_shared"), path.join(TMP, "_shared"));
await cp(path.join(REPO, "demo-brutalist"), path.join(TMP, "demo-brutalist"));
await cp(path.join(REPO, "demo-moonlight"), path.join(TMP, "demo-moonlight"));
await fs.copyFile(path.join(REPO, "_quarto.yml"), path.join(TMP, "_quarto.yml"));
console.log("temp root:", TMP);

// Pull in the modules from the real app/ (so we get the real implementation)
const { listAllSlides, getDeck } = await import(url.pathToFileURL(path.join(REPO, "app/server/decks.js")).href);
const { addMissingStyles } = await import(url.pathToFileURL(path.join(REPO, "app/server/style-copy.js")).href);

const slides = await listAllSlides(TMP);
const splitSlide = slides.find((s) => s.deckId === "demo-brutalist" && s.title === "SPLIT");
if (!splitSlide) throw new Error("SPLIT slide not found");

console.log("\nsource slide raw block (first 400 chars):");
console.log(splitSlide.rawBlock.slice(0, 400));

const sourceDeck = await getDeck(TMP, "demo-brutalist");
const targetDeck = await getDeck(TMP, "demo-moonlight");

const beforeCss = await fs.readFile(path.join(TMP, "demo-moonlight/moonlight.css"), "utf8");
const result = await addMissingStyles(TMP, sourceDeck, targetDeck, splitSlide.rawBlock, "missing");
const afterCss = await fs.readFile(path.join(TMP, "demo-moonlight/moonlight.css"), "utf8");

console.log("\n=== MISSING-ONLY MODE ===");
console.log("result:", JSON.stringify(result, null, 2));
console.log("---- appended ----");
console.log(afterCss.slice(beforeCss.length));

// Re-run missing-only: should add nothing.
const again = await addMissingStyles(TMP, sourceDeck, targetDeck, splitSlide.rawBlock, "missing");
console.log("\nsecond missing run:", JSON.stringify(again, null, 2));

// Restore moonlight.css and try override mode.
await fs.writeFile(path.join(TMP, "demo-moonlight/moonlight.css"), beforeCss);
const override = await addMissingStyles(TMP, sourceDeck, targetDeck, splitSlide.rawBlock, "override");
const overrideCss = await fs.readFile(path.join(TMP, "demo-moonlight/moonlight.css"), "utf8");
console.log("\n\n=== OVERRIDE MODE ===");
console.log("added:", override.added.length, "replaced:", override.replaced.length,
            "vars+:", override.vars?.length, "vars~:", override.varsReplaced?.length);
console.log("file size before/after:", beforeCss.length, "->", overrideCss.length);

// Spot-check: target had `.reveal h2 { ... }` originally. Did the source's
// version replace it in-place (not appended)?
const h2BeforeIdx = beforeCss.indexOf(".reveal h2");
const h2AfterIdx = overrideCss.indexOf(".reveal h2");
console.log("`.reveal h2` index:", h2BeforeIdx, "->", h2AfterIdx,
            "(in-place if both > 0 and close)");

// Re-run override: should replace what's there, add nothing new.
const override2 = await addMissingStyles(TMP, sourceDeck, targetDeck, splitSlide.rawBlock, "override");
console.log("second override run:", "added:", override2.added.length, "replaced:", override2.replaced.length);

File: _tmp/test_style_copy.py
"""End-to-end smoke test for cross-deck paste + auto CSS copy.

Spins up the app server against a temp copy of two decks, posts a paste action
from one deck into the other, and reports what got added to the target CSS.
"""
import json
import os
import shutil
import subprocess
import sys
import tempfile
import time
import urllib.request
from pathlib import Path

REPO = Path(r"C:\Users\Zoom\My Drive (hello@causalmap.app)\Causal Map\10-19 Outreach - marketing - presentations - academic - theory\19c-slides")
APP = REPO / "app"


def http(method, path, body=None, port=3298):
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(
        f"http://127.0.0.1:{port}{path}",
        data=data,
        method=method,
        headers={"content-type": "application/json"} if body is not None else {},
    )
    with urllib.request.urlopen(req, timeout=15) as r:
        return json.loads(r.read())


def main():
    # Use an isolated temp root so we don't mutate Steve's files
    tmp = Path(tempfile.mkdtemp(prefix="slides-test-"))
    shutil.copytree(REPO / "_shared", tmp / "_shared")
    shutil.copytree(REPO / "demo-brutalist", tmp / "demo-brutalist")
    shutil.copytree(REPO / "demo-moonlight", tmp / "demo-moonlight")
    shutil.copy(REPO / "_quarto.yml", tmp / "_quarto.yml")
    print("temp root:", tmp)

    env = {
        **os.environ,
        "PORT": "3298",
    }
    # Server resolves repoRoot as parent of app/. Move app/ into tmp.
    shutil.copytree(APP, tmp / "app", ignore=shutil.ignore_patterns("node_modules", "dist", ".vite"))

    p = subprocess.Popen(["node", str(tmp / "app" / "server" / "index.js")], env=env,
                         cwd=str(tmp / "app"),
                         stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    try:
        time.sleep(2)
        # Pick a brutalist slide that uses several classes
        slides = http("GET", "/api/slides/all")
        bru = [s for s in slides if s["deckId"] == "demo-brutalist"]
        # Find SPLIT slide (uses .splitgrid .panel-black .panel-yellow .label etc)
        target_slide = next((s for s in bru if s["title"] == "SPLIT"), bru[3])
        print("source slide:", target_slide["deckId"], target_slide["title"])

        # Capture target CSS before
        moon_css = (tmp / "demo-moonlight" / "moonlight.css").read_text()
        before_len = len(moon_css)

        # Pick anything in moonlight to insert after
        moon_slides = [s for s in slides if s["deckId"] == "demo-moonlight"]
        moon_after = moon_slides[2]["slideId"]

        resp = http("POST", "/api/decks/demo-moonlight/slides/apply", {
            "type": "paste",
            "afterSlideId": moon_after,
            "blocks": [target_slide["rawBlock"]],
            "sourceDeckId": "demo-brutalist",
        })

        print("stylesAdded:", json.dumps(resp.get("stylesAdded"), indent=2))

        after = (tmp / "demo-moonlight" / "moonlight.css").read_text()
        added = after[before_len:]
        print("---- appended to moonlight.css ----")
        print(added)
        print("-----------------------------------")
    finally:
        p.terminate()
        try: p.wait(timeout=3)
        except Exception: p.kill()


if __name__ == "__main__":
    main()
